### ISSUE TITLE
Review installation logic, files and directories

### ISSUE THREAD

**Issue by zbeekman** [sourceryinstitute/opencoarrays#79]
- Add an `uninstall` target so users can easily remove OCA (now #89)
- Investigate which files are being installed, to where, and why
  - right now it seems that some of the tests are being installed; these should probably only be run at build time `cmake .. && make && make test && make install` is the canonical order of operations... no need to install tests.
- Fortran and .mod files complicate this a little bit, when on systems with multiple fortran compilers. I developed what I think is a very good system for https://github.com/jacobwilliams/JSON-Fortran

@zbeekman I'm adding an item to this list:
- Write unit tests for the installation scripts.  

I started this with a [commit](https://github.com/sourceryinstitute/opencoarrays/commit/91f8fd1cc1ac3373b1ab6a46d91aad6c25c2d492) pushed last night.  The motivation was that my work on inserting `find_or_install gcc` near the beginning of [build_opencoarrays.sh](https://github.com/sourceryinstitute/opencoarrays/blob/master/prerequisites/install-functions/build_opencoarrays.sh) necessitates changing some stack management in [find_or_install.sh](https://github.com/sourceryinstitute/opencoarrays/blob/master/prerequisites/install-functions/find_or_install.sh) and the stack function I'll most likely use is one that I wrote, but hadn't sufficiently tested. The new tests have already identified some issues in stack.sh and helped define the boundaries of what is safe to use.

**Comment by rouson** [comment]
* Handle additional missing prerequisites: wget, svn, and m4. 
  * When wget and/or svn is missing, print a URL for each file that needs to be downloaded and direct the user download the file via a web browser.
  * When m4 is missing, attempt to install it.

**Comment by zbeekman** [comment]
Just curious, has missing wget, and svn been handled yet? I am expanding the test matrix to cover the install script.

**Comment by rouson** [comment]
@zbeekman, I'm not sure what you mean but yes, that's the subject of my most recent commit merged into master.

**Comment by zbeekman** [comment]
Ah yes, I thought I remembered that this was added, but perhaps I need to pull in the latest commits from master to my branch. Thanks for clarifying.

**Comment by zbeekman** [comment]
@rouson does d5783167707dfb8d2dc973414fa14cd483c57a20 handle both svn/subversion *and* wget? (I looked it over but it's quite sophisticated, and will take some time to fully digest)

**Comment by rouson** [comment]
@zbeekman, short answer: yes.  

Long answer: wget, svn, and ftp are all invoked by evaluating a variable named "flex", which is set to one of the aforementioned values ("wget", "svn", or "ftp") inside install_prerequisites/build.  Commit #100 address the case when "type $fetch" exits with a non-zero value indicating the command referenced by $fetch wasn't found.  This should work for all of the above (wget, svn, or ftp), but I only tested it with wget because that's the only one anybody complained about and because wget is the most convenient one for me to temporarily uninstall for testing purposes. (I install wget via MacPorts, whereas svn and ftp are part of OS X.)   The fix will work just as well for svn and ftp because all it does is let the user know the the relevant program is missing and give the user a URL to download the file that the script would have downloaded via wget/ftp/svn.  The only issue is that the message that prints might need some tweaking because the string that I expect to be the URL might not have the right format in the ftp/svn case.  It wouldn't take a ton of effort for me to figure out if the latter sentence is true.  I just haven't take the time to do it, but the worst that can happen is that the user sees a message the might not be as helpful as it could be.

**Comment by zbeekman** [comment]
Ah, I see. So the script does not attempt to install any of the `fetch` method software if it's missing because it doesn't know how to download it; rather it points the user in the direction of where to obtain the software in question that it needs.

Thanks.

**Comment by rouson** [comment]
@zbeekman, install_prerequisites/build contains a variable named "fetch" that can be set to one of several values: "wget", "svn", or "ftp".  The latest commit handles the case when the value of fetch .  It's just that I've only tested it in the case when the value of wget is missing.

**Comment by zbeekman** [comment]
@rouson :+1:

**Comment by rouson** [comment]
Earlier in this thread, I mentioned handling missing wget, svn, and m4.   Now that we have a workaround for wget and svn, only m4 remains.   The [build](https://github.com/sourceryinstitute/opencoarrays/blob/master/install_prerequisites/build) script is already capable of building and installing m4. Once I figure out where to insert m4 in the [dependency tree](https://github.com/sourceryinstitute/opencoarrays/blob/master/doc/dependency_tree/opencoarrays-tree.txt), modifying [install.sh](https://github.com/sourceryinstitute/opencoarrays/blob/master/install.sh) will be straightforward. 

MacPorts to the rescue (sort of): "sudo port uninstall m4" shows me that bison depends on m4.  bison is currently at the tail end of the gcc branch of our dependency tree so I'll append m4 there in a new branch.   

FYI, Cray engineer Harvey Richardson tried the useful exercise of building OpenCoarrays in a bare Linux vm that he had just downloaded and booted without modification.  From that experience, which he presented in [slides](http://www.fortran.bcs.org/2015/BCS_FSG_2015_HR.pdf) at a meeting of the British Computer Society Fortran Specialist Group, he identified m4 as a prerequisite, but until this moment, I never knew where to insert it in the dependency tree and I didn't want to go with the unnecessary approach of always installing it when it's missing.  Mystery solved!  

I'll branch from master and see if I can add the logic for installing m4 when necessary inside [install.sh](https://github.com/sourceryinstitute/opencoarrays/blob/master/install.sh).  To test, I might not be hard to temporarily fake the script into thinking m4 is missing.

**Comment by rouson** [comment]
The m4 dependency is now handed by #103.

**Comment by zbeekman** [comment]
I would like to sneak this into the next release if I can swing it.

**Comment by rouson** [comment]
:+1: as long as it happens soon.

**Comment by zbeekman** [comment]
OK, unless I'm mistaken it looks like all the prerequisite stuff @rouson added/was talking about has been taken care of. I'm going to examine where everything gets put and what needs to happen to enable CMake's `find_package(OpenCoarrays)`

**Comment by zbeekman** [comment]
`install_manifest.txt`:

```
/usr/local/lib/cmake/opencoarrays/OpenCoarraysTargets.cmake
/usr/local/lib/cmake/opencoarrays/OpenCoarraysTargets-release.cmake
/usr/local/lib/cmake/opencoarrays/OpenCoarraysConfig.cmake
/usr/local/lib/cmake/opencoarrays/OpenCoarraysConfigVersion.cmake
/usr/local/./mod/opencoarrays.mod
/usr/local/lib/libcaf_mpi.a
/usr/local/./mod/opencoarrays.mod
/usr/local/bin/caf
/usr/local/bin/cafrun
```

Here is a lengthy discussion about what to do with the `.mod` files, which is the only think that may be worth changing: https://github.com/jacobwilliams/json-fortran/issues/98

**Comment by zbeekman** [comment]
Also, add a `opencoarrays.pc` file, see https://github.com/jacobwilliams/json-fortran/pull/242 for reference implementation

**Comment by rouson** [comment]
Cool, but I'd say the most important thing at this point is to just to release ASAP -- hopefully no later than next Tuesday sometime after our weekly call.  From the [agile manifesto](http://agilemanifesto.org/principles.html):

"Deliver working software frequently, from a couple of weeks to a couple of months, with a 
preference to the shorter timescale."

It's more important to release frequently than to try to squeeze additional features into the next release. There's nothing wrong with releasing again soon to get the new features out.  In fact, releasing again soon is part of the objective.

**Comment by zbeekman** [comment]
Also, the header file is noticeably missing from the installation manifest...
